tests/test_group.py

```diff
@@ -0,0 +1,139 @@
+        q.empty()
+
+    def test_fetch_group(self):
+        q = Queue(connection=self.testconn)
+        enqueued_group = Group.create(connection=self.testconn)
+        enqueued_group.enqueue_many(q, [self.job_1_data, self.job_2_data])
+        fetched_group = Group.fetch(enqueued_group.id, self.testconn)
+        self.assertCountEqual(enqueued_group.get_jobs(), fetched_group.get_jobs())
+        assert len(fetched_group.get_jobs()) == 2
+        q.empty()
+
+    def test_add_jobs(self):
+        q = Queue(connection=self.testconn)
+        group = Group.create(connection=self.testconn)
+        group.enqueue_many(q, [self.job_1_data, self.job_2_data])
+        job2 = group.enqueue_many(q, [self.job_1_data, self.job_2_data])[0]
+        assert job2 in group.get_jobs()
+        self.assertEqual(job2.group_id, group.id)
+        q.empty()
+
+    def test_jobs_added_to_group_key(self):
+        q = Queue(connection=self.testconn)
+        group = Group.create(connection=self.testconn)
+        jobs = group.enqueue_many(q, [self.job_1_data, self.job_2_data])
+        job_ids = [job.id for job in group.get_jobs()]
+        jobs = list({as_text(job) for job in self.testconn.smembers(group.key)})
+        self.assertCountEqual(jobs, job_ids)
+        q.empty()
+
+    def test_group_id_added_to_jobs(self):
+        q = Queue(connection=self.testconn)
+        group = Group.create(connection=self.testconn)
+        jobs = group.enqueue_many(q, [self.job_1_data])
+        assert jobs[0].group_id == group.id
+        fetched_job = Job.fetch(jobs[0].id, connection=self.testconn)
+        assert fetched_job.group_id == group.id
+
+    def test_deleted_jobs_removed_from_group(self):
+        q = Queue(connection=self.testconn)
+        group = Group.create(connection=self.testconn)
+        group.enqueue_many(q, [self.job_1_data, self.job_2_data])
+        job = group.get_jobs()[0]
+        job.delete()
+        group.cleanup()
+        redis_jobs = list({as_text(job) for job in self.testconn.smembers(group.key)})
+        assert job.id not in redis_jobs
+        assert job not in group.get_jobs()
+
+    def test_group_added_to_registry(self):
+        q = Queue(connection=self.testconn)
+        group = Group.create(connection=self.testconn)
+        group.enqueue_many(q, [self.job_1_data])
+        redis_groups = {as_text(group) for group in self.testconn.smembers("rq:groups")}
+        assert group.id in redis_groups
+        q.empty()
+
+    def test_expired_jobs_removed_from_group(self):
+        q = Queue(connection=self.testconn)
+        w = SimpleWorker([q], connection=q.connection)
+        short_lived_job = Queue.prepare_data(say_hello, result_ttl=1)
+        group = Group.create(connection=self.testconn)
+        group.enqueue_many(q, [short_lived_job, self.job_1_data])
+        w.work(burst=True, max_jobs=1)
+        sleep(3)
+        group.cleanup()
+        assert len(group.get_jobs()) == 1
```

Comment:
Can we assert that it's `job_1_data` that remains in the group?

Reply:
Done